File: backend/app/schemas/treasury.py
from datetime import date, datetime
from decimal import Decimal
from enum import Enum

from pydantic import BaseModel, ConfigDict, Field


class TreasuryType(str, Enum):
    BILL = "BILL"
    NOTE = "NOTE"
    BOND = "BOND"


class TreasuryBuyRequest(BaseModel):
    """
    purchase_price and (later) sale_price are quoted per 100 of face value,
    matching how Treasury prices are dealt/displayed (e.g. 98.50 means $985
    per $1,000 face). coupon_frequency is payments per year — 2 for
    semi-annual notes/bonds, 0 for a zero-coupon bill.
    """

    treasury_type: TreasuryType
    face_value: Decimal = Field(gt=0)
    coupon_rate: Decimal = Field(ge=0)
    coupon_frequency: int = Field(default=2, ge=0, le=12)
    maturity_date: date
    purchase_price: Decimal = Field(gt=0)
    purchase_date: date | None = None


class TreasurySellRequest(BaseModel):
    sale_price: Decimal = Field(gt=0)


class TreasuryHoldingResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    treasury_type: TreasuryType
    face_value: Decimal
    coupon_rate: Decimal
    coupon_frequency: int
    maturity_date: date
    purchase_date: date
    purchase_price: Decimal


class TreasuryValuationResponse(TreasuryHoldingResponse):
    current_price: Decimal
    cost_basis: Decimal
    market_value: Decimal
    gain_loss: Decimal
    gain_loss_percentage: Decimal


class YieldCurvePoint(BaseModel):
    tenor_years: float
    yield_percent: Decimal


class YieldCurveResponse(BaseModel):
    as_of: datetime | None
    points: list[YieldCurvePoint]


class TreasurySellResponse(BaseModel):
    holding_id: int
    ticker: str
    proceeds: Decimal
    cost_basis: Decimal
    realized_gain_loss: Decimal
    cash_balance_after: Decimal

File: frontend/src/types/treasury.ts
/**
 * Types mirroring backend/app/schemas/treasury.py
 */

export type TreasuryType = "BILL" | "NOTE" | "BOND";

/** Payload for POST /treasury/buy */
export type TreasuryBuyRequest = {
  treasury_type: TreasuryType;
  face_value: string;
  coupon_rate: string;
  coupon_frequency: number;
  maturity_date: string;
  purchase_price: string;
  purchase_date?: string | null;
};

/** Payload for POST /treasury/:id/sell */
export type TreasurySellRequest = {
  sale_price: string;
};

export type TreasuryHolding = {
  id: number;
  treasury_type: TreasuryType;
  face_value: string;
  coupon_rate: string;
  coupon_frequency: number;
  maturity_date: string;
  purchase_date: string;
  purchase_price: string;
};

export type TreasuryValuation = TreasuryHolding & {
  current_price: string;
  cost_basis: string;
  market_value: string;
  gain_loss: string;
  gain_loss_percentage: string;
};

export type YieldCurvePoint = {
  tenor_years: number;
  yield_percent: string;
};

export type YieldCurve = {
  as_of: string | null;
  points: YieldCurvePoint[];
};

export type TreasurySellResult = {
  holding_id: number;
  ticker: string;
  proceeds: string;
  cost_basis: string;
  realized_gain_loss: string;
  cash_balance_after: string;
};

/** Light multi-asset seam for dashboard allocation. */
export type AssetClassId = "stock" | "treasury";
